src/Compilers/CSharp/Portable/FlowAnalysis/VariableUsePass.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// VariableUsePass marks all referenced variables in the visited node as used (as a read).
+    /// This pass is used for things like attribute arguments and default parameter expressions,
+    /// which are otherwise not analyzed for use references.
+    /// </summary>
+    internal sealed class VariableUsePass : BoundTreeWalkerWithStackGuard
+    {
+        /// <summary>
+        /// The current source assembly.
+        /// </summary>
+        private readonly SourceAssemblySymbol _sourceAssembly;
+
+        /// <summary>
+        /// Variables that were used anywhere, in the sense required to suppress warnings about
+        /// unused variables.
+        /// </summary>
+        private readonly HashSet<LocalSymbol> _usedVariablesOpt;
+
+        /// <summary>
+        /// Variables that were used anywhere, in the sense required to suppress warnings about
+        /// unused variables.
+        /// </summary>
+        private readonly HashSet<LocalFunctionSymbol> _usedLocalFunctionsOpt;
+
+        /// <summary>
+        /// Create a stateless visitor (except for the SourceAssembly).
+        /// </summary>
+        /// <remarks>
+        /// An instance of this class is cached on CSharpCompilation, so be careful adding new state.
+        /// (Either avoid state that cannot be shared across the compilation, or do not cache an instance on the compilation)
+        /// </remarks>
+        public VariableUsePass(SourceAssemblySymbol sourceAssembly)
+            : this(sourceAssembly, null, null)
+        {
+        }
+
+        /// <summary>
+        /// Create a visitor for use within a method (e.g. local functions).
+        /// Keep track of what local variables are referenced, by placing them in the HashSets.
+        /// </summary>
+        public VariableUsePass(SourceAssemblySymbol sourceAssembly, HashSet<LocalSymbol> usedVariables, HashSet<LocalFunctionSymbol> usedLocalFunctions)
+        {
+            _sourceAssembly = sourceAssembly;
+            _usedVariablesOpt = usedVariables;
+            _usedLocalFunctionsOpt = usedLocalFunctions;
+        }
+
+        public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+        {
+            _sourceAssembly.NoteFieldAccess(node.FieldSymbol, true, false);
```

Comment:
>_sourceAssembly.NoteFieldAccess(node.FieldSymbol, true, false); [](start = 12, length = 63)

It looks like in order to have desired effect, this method should be called on a definition of the field. Please add a test to make this problem observable, then fix it. It looks like several more call-sites of this API have the same problem, we should probably change implementation of NoteFieldAccess to dig to the definition.